compiler/grainlsp/rpc.re

```diff
@@ -0,0 +1,424 @@
+type diagnostics_message = {
+  jsonrpc: string,
+  method: string,
+  params: document_diagnostics,
+};
+
+[@deriving yojson]
+type markup_content = {
+  kind: string,
+  value: string,
+};
+
+[@deriving yojson]
+type marked_string = {
+  language: string,
+  value: string,
+};
+
+[@deriving yojson]
+type hover_result = {
+  contents: markup_content,
+  range,
+};
+
+[@deriving yojson]
+type hover_response = {
+  jsonrpc: string,
+  id: int,
+  result: hover_result,
+};
+
+[@deriving yojson]
+type definition_result = {
+  uri: string,
+  range,
+};
+
+[@deriving yojson]
+type definition_response = {
+  jsonrpc: string,
+  id: int,
+  result: definition_result,
+};
+
+[@deriving yojson]
+type completion_item = {
+  label: string,
+  kind: int,
+  detail: string,
+  documentation: string,
+};
+
+[@deriving yojson]
+type completion_result = {
+  isIncomplete: bool,
+  items: list(completion_item),
+};
+
+[@deriving yojson]
+type completion_response = {
+  jsonrpc: string,
+  id: int,
+  result: completion_result,
+};
+
+let convert_range = range => {
+  let r_start: position = {
+    line: range.start_line - 1,
+    character: range.start_char,
+  };
+  let r_end: position = {line: range.end_line - 1, character: range.end_char};
+  {start: r_start, range_end: r_end};
+};
+
+let read_message = (input): protocol_msg => {
+  let clength = input_line(input);
+  let cl = "Content-Length: ";
+  let cll = String.length(cl);
+  if (String.sub(clength, 0, cll) == cl) {
+    /* if on windows, dont need the extra -1 */
+    let offset = Sys.os_type == "Win32" ? 0 : (-1); /* -1 for trailing \r */
+
+    let num =
+      String.sub(clength, cll, String.length(clength) - cll + offset);
+    let num = (num |> int_of_string) + (Sys.os_type == "Win32" ? 1 : 2);
+    let buffer = Buffer.create(num);
+    Buffer.add_channel(buffer, input, num);
+    let raw = Buffer.contents(buffer);
+    let json = Yojson.Safe.from_string(raw);
+
+    let action =
+      Yojson.Safe.Util.member("method", json) |> Yojson.Safe.Util.to_string;
+
+    let idOpt =
+      Yojson.Safe.Util.member("id", json) |> Yojson.Safe.Util.to_int_option;
+
+    switch (idOpt) {
+    | None => Notification(action, json)
+    | Some(id) => Message(id, action, json)
+    };
+  } else {
+    failwith("Invalid header");
+  };
+};
+
+let send = (output, content) => {
+  let length = String.length(content);
+  let sep = Sys.os_type == "Unix" ? "\r\n\r\n" : "\n\n";
+  let len = string_of_int(length);
+
+  output_string(output, "Content-Length: " ++ len ++ sep ++ content);
+  flush(output);
+};
+
+let send_null_message = (output, id) => {
+  let res =
+    `Assoc([
+      ("jsonrpc", `String("2.0")),
+      ("id", `Int(id)),
+      ("result", `Null),
+    ]);
+
+  let strJson = Yojson.Safe.pretty_to_string(res);
+  send(output, strJson);
+};
+
+let send_capabilities = (output, id: int) => {
+  let completion_vals: completion_values = {
+    resolveProvider: true,
+    triggerCharacters: ["."],
```

Comment:
Missed these on the definitions, but no camelcase please. I believe you can do yojson mappings to convert between camelcase and snakecase

Reply:
I'm going to push back on this :). These are only used within this file and are the LSP protocol fields.  To me it makes it clearer to future developers where this stuff is set and how it maps, rather than to a yojson mapping for the sake of it.   
If this record was accessed outside this file I'd map them.  It's almost like a lowlevel implementation that should have @unsafecamelcase tagged on it ;-)

Reply:
All fixed by @phated !